### PR TITLE
MIDI4lights: Give beginTimer callbacks the anonymous function expression form

### PR DESCRIPTION
**Pull request by geraldog**
Merged PR #11953 "fix(controllers): engine.beginTimer(ms, string) would evaluate the code immediately instead of after ms" requires anonymous functions expressions instead of string literals.

This should solve #12042

### CODE REVIEW

**Comment by Swiftb0y** [comment]
https://github.com/mixxxdj/mixxx/pull/11953 should've fixed this exact issue already though. If #11953 works correctly, this PR would just be noop. Do you think #11953 is broken?

**Comment by geraldog** [comment]
Hi @Swiftb0y and sorry for bringing you out of your time off.

I tested again and again with real fixture to make sure we understand better the problem and save some face. Still, it seems I need to test even more but let me explain.

I use a modified midi_for_light.vuMeter to interact with AudioPathType::Stem I have created in my fourstems_padj_2023 branch.

Without the proposed fix, here's the warning I get when I engage my Programmable AutoDJ with Stems capabilities:

` controller.midithroughport0:warning [Controller] "passed a string to `engine.beginTimer`, please pass a function instead!"`

And it doesn't work properly. I specifically lose the on-beat information!

**Comment by geraldog** [comment]
@Swiftb0y as not to waste your precious time I reproduced #12042 with Vanilla Mixxx (no weird customizations). I had a hunch the author of #12042 was using AutoDJ too like me, and lo and behold if you don't touch the crossfader but instead rely on AutoDJ to do crossfading to the next song for you, the bug happens and I get the same warning as with the strange customizations.

**Comment by Swiftb0y** [comment]
I appreciate it highly btw that you are so understanding of my personal resource constraints. Thank you.

**Comment by geraldog** [comment]
> I have no doubt that you are experiencing the issue you're describing (I have never looked into the midi4lights script myself so I'm not the right person for this anyways). The thing I don't understand is how this PR is supposed to fix #12042. Since as I said, the behavior should I identical (assuming #11953 works as intended).
> 
> So, are you confident that this PR actually fixes #12042? I'm unfortunately not able to confirm for myself.

No, I'm not confident at all. We need the author of #12042 to test the proposed fix, see if it solves his bug too.

File: res/controllers/Midi_for_light-scripts.js [hunk 1]
```diff
@@ -416,7 +416,7 @@ midi_for_light.deckVolumeChange = function(value, group, control) { // deck volu
         midi_for_light.deck_current = deckneu;
         midi.sendShortMsg(0x8F + midi_channel, 0x30, 0x64 + deckneu); // Note C on with 64 and add deck
         midi_for_light.volumeBeatBlockStatus = true;
-        midi_for_light.volumeBeatBlock_timer = engine.beginTimer(1000, "midi_for_light.volumeBeatBlock()");
+        midi_for_light.volumeBeatBlock_timer = engine.beginTimer(1000, function() { midi_for_light.volumeBeatBlock(); });
```
Comment (Holzhaus) [inline res/controllers/Midi_for_light-scripts.js:419]: Does this work, too? 

```suggestion
        midi_for_light.volumeBeatBlock_timer = engine.beginTimer(1000, midi_for_light.volumeBeatBlock);
```

File: res/controllers/Midi_for_light-scripts.js [hunk 2]
```diff
@@ -441,7 +441,7 @@ midi_for_light.crossfaderChange = function() { // crossfader chenge, check deck
     engine.stopTimer(midi_for_light.volumebeat_on_delay_timer);
     if (engine.getValue("[Master]", "crossfader") > -0.25) { // crossfader more than 25% left;
         if (engine.getValue("[Master]", "crossfader") < 0.25) { // crossfader more then 25% right;
-            midi_for_light.volumebeat_on_delay_timer = engine.beginTimer(3000, "midi_for_light.volumeBeatOnDelay()");
+            midi_for_light.volumebeat_on_delay_timer = engine.beginTimer(3000, function() { midi_for_light.volumeBeatOnDelay(); });
```
Comment (Holzhaus) [inline res/controllers/Midi_for_light-scripts.js:444]:
```suggestion
            midi_for_light.volumebeat_on_delay_timer = engine.beginTimer(3000, midi_for_light.volumeBeatOnDelay);
```
?

File: res/controllers/Midi_for_light-scripts.js [hunk 3]
```diff
@@ -463,7 +463,7 @@ midi_for_light.crossfaderChange = function() { // crossfader chenge, check deck
         midi_for_light.deck_current = deck;
         midi.sendShortMsg(0x8F + midi_channel, 0x30, 0x64 + deck); // note C on with value 64 + deck
         midi_for_light.crossfader_block = true;
-        midi_for_light.crossfader_change_block_timer = engine.beginTimer(1000, "midi_for_light.crossfaderChangeBlock()");
+        midi_for_light.crossfader_change_block_timer = engine.beginTimer(1000, function() { midi_for_light.crossfaderChangeBlock(); });
```
Comment (Holzhaus) [inline res/controllers/Midi_for_light-scripts.js:466]:
```suggestion
        midi_for_light.crossfader_change_block_timer = engine.beginTimer(1000, midi_for_light.crossfaderChangeBlock);
```

**Review by Holzhaus — COMMENTED** [review]
Thanks. Please check if we can pass the function directly instead of wrapping it in an anonymous function.

**Comment by geraldog** [comment]
Yes Jan it works normally by passing the function directly, except for the functions which need arguments of course, which I kept as anonymous function expressions.

**Comment by Holzhaus** [comment]
> Yes Jan it works normally by passing the function directly, except for the functions which need arguments of course, which I kept as anonymous function expressions.

Good :+1:

I think you didn't push your changes yet though.

**Review by Holzhaus — APPROVED** [review]
Code changes look good to make and make sense, even if they don't fix #12042 (the changes should make no difference, but I didn't test).

@geraldog it would be great if you could apply the proposed fixes from eslint to fix the pre-commit CI failure, too.

**Comment by geraldog** [comment]
Hi Jan, thanks for the review by the way, and thank you @ronso0 for updating with the correct title, my bad.

> Code changes look good to make and make sense, even if they don't fix #12042 (the changes should make no difference, but I didn't test).

The changes do make a difference Jan, they don't go through that code path that Niko fixed inside src/controllers/scripting/legacy/controllerscriptinterfacelegacy.cpp in #11953. Namely:

`int ControllerScriptInterfaceLegacy::beginTimer(
        int intervalMillis, QJSValue timerCallback, bool oneShot) {
    if (timerCallback.isString()) {
        m_pScriptEngineLegacy->logOrThrowError(
                QStringLiteral("passed a string to 'engine.beginTimer', please "
                               "pass a function instead!"));
        // wrap the code in a function to make the evaluation lazy.
        // otherwise the code would be evaluated immediately instead of after
        // the timer which is obviously undesired and could also cause
        // issues when used recursively.
        timerCallback = m_pScriptEngineLegacy->jsEngine()->evaluate(
                QStringLiteral("()=>%1").arg(timerCallback.toString()));
    } else if (!timerCallback.isCallable()) {`

> @geraldog it would be great if you could apply the proposed fixes from eslint to fix the pre-commit CI failure, too.

I will look into it Jan, and I don't want to be rude that's why I held back on this comment I'm about to make.

Niko said the script needs more love, but I consider this an understatement. Quite frankly that script looks like one of my hacks, amazing but kind-of constructed in a rush, if you know what I mean.

**Review by Swiftb0y — APPROVED** [review]
welp, lgtm, eslint failure is expected. Still very likely _won't_ solve https://github.com/mixxxdj/mixxx/issues/12042

**Comment by geraldog** [comment]
Thanks @Swiftb0y

**Pull request merged**